docs/proposals/webhook-validation.md

```diff
@@ -0,0 +1,130 @@
+
+- Include a mutating or defaulting or conversion webhook.
+  If we decide these are useful, we can add them in future proposals.
+
+## Proposal
+
+Rely on [kubebuilder's webhook support](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html) to add a new webhook validation controller, and include it in Shipwright's default installation as a K8s Deployment.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Certificates 
+
+The webhook deployment will also need to manage certificates since K8s webhook requests are sent over HTTPS.
+Setting up these certificates securely can be cumbersome.
+
+The [Knative](https://knative.dev) project has [some libraries to make this management easier](https://github.com/knative/pkg/tree/main/webhook) by taking care of cert generation, registration and refreshing, but mixing Knative-style components and existing `controller-gen`/KubeBuilder-style components might make complicate development.
+
+#### Webhook Availability
+
+Once registered, if the K8s API server can't reach the defined webhook service, _all_ requests will fail for registered CRD types.
+The webhook deployment is stateless, so it can be easily replicated and scheduled across multiple nodes to ensure availability in the case of node failure.
+
+### Risks and Mitigations
+
+Clients that expect to have requests accepted and asynchronously validated might be surprised to find the API synchronously rejecting their invalid requests instead.
+We should communicate this change to users and operators in release notes and in mailing list announcements, and be available to answer any questions.
+
+## Design Details
+
+### Test Plan
+
+Our current tests to validate resources should be mostly reusable; validation logic itself isn't going anywhere, just moving.
+
+Any tests that we have that assume an invalid resource will be accepted and eventually reconciled should be updated to expect immediate failure.
+
+### Version Skew Strategy
+
+A user who has previously installed a version of Shipwright that did not synchronously validate might have invalid resources in their etcd storage, and our controller may be asked to reconcile them.
+Basic validation logic should remain in the reconciler, but either update a status to indicate invalidity, or merely log and ignore invalid resources.
+
+During installation or upgrade of Shipwright components, the webhook job and the reconciling controller job might briefly run two versions of validation code, so we should be careful to consider this as we add (or remove, or refactor) validation support over time.
+In practice, Tekton hasn't experienced significant issues with this, but I wanted to call it out because it's possible.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation History`.
+
+## Drawbacks
+
+This requires us to build a new component in the Shipwright installation, and for the operator component to operate it.
```

Comment:
That list is quite short. :-)

There is one caveat with the webhook: it validates only on API request (create or update is relevant in this context). But, our validations often include another resource which can also change.

We have some validations where this does not really matter, for example the Git repository validation because we have no trigger if this goes away anyway, or the non-existing Build on a BuildRun creation because the BuildRun is a one-shot thing.

But, we also validate the referenced secrets of a Build and would notice if a secret that is referenced by a Build gets deleted (assuming the secret has the necessary annotation) and subsequently mark the Build as not registered. With the webhook, this would not be possible anymore.

Reply:
Good points, thanks for bringing them up.

I think it could be simpler (for our implementation, and our users' mental model) for Builds to be assumed to be valid until we find out they're not the next time they're run. I'd like to avoid the scenario where deleting/updating a Secret results in dozens or hundreds of Builds having to be reconciled to check if they're still valid. Especially if that validation includes reaching out to a registry to check that the new credentials are valid, potentially lots of times.

In Tekton, you can create a Task, create a Pipeline that uses the Task, then delete the Task, and the Pipeline doesn't proactively report as invalid. The next time you try to run that Pipeline, PipelineRun creation will fail saying the referenced Task isn't valid. This avoids unnecessary duplicative validation work while a user doesn't care, and puts the error message directly in the user's face the next time they try it.

I'm open to discussing this point more though.

Reply:
Sounds good.